ros/angel_system_nodes/angel_system_nodes/audio_player.py

```diff
@@ -0,0 +1,108 @@
+import threading
+import time
+
+import simpleaudio as sa
+
+import rclpy
+from rclpy.node import Node
+from angel_msgs.msg import HeadsetAudioData
+
+
+class AudioPlayer(Node):
+    def __init__(self):
+        super().__init__(self.__class__.__name__)
+
+        self.declare_parameter("audio_topic", "HeadsetAudioData")
+
+        self._audio_topic = self.get_parameter("audio_topic").get_parameter_value().string_value
+
+        self.subscription = self.create_subscription(
+            HeadsetAudioData,
+            self._audio_topic,
+            self.listener_callback,
+            1)
+
+        self.audio_stream = bytearray()
+        self.audio_stream_lock = threading.RLock()
+        self.audio_player_lock = threading.RLock()
+        self.t = threading.Thread()
+
+        self.audio_stream_duration = 0.0
+        self.playback_duration = 1.0
+
+        self.prev_timestamp = None
+
+    def listener_callback(self, msg):
+        """
+        Add the audio to the cumulative audio stream and spawns a new playback
+        thread if the cumulative recording has reached more than a second and
+        there is no ongoing playback.
+        """
+        log = self.get_logger()
+
+        with self.audio_stream_lock:
+            # Check that this message comes temporally after the previous message
+            message_order_valid = False
+            if self.prev_timestamp is None:
+                message_order_valid = True
+            else:
+                if msg.header.stamp.sec > self.prev_timestamp.sec:
+                    message_order_valid = True
+                elif msg.header.stamp.sec == self.prev_timestamp.sec:
+                    # Seconds are same, so check nanoseconds
+                    if msg.header.stamp.nanosec > self.prev_timestamp.nanosec:
+                        message_order_valid = True
+
+            if message_order_valid:
+                self.audio_stream.extend(msg.data)
+                self.audio_stream_duration += msg.sample_duration
+            else:
+                log.info("Warning! Out of order messages.\n"
+                         + f"Prev: {self.prev_timestamp} \nCurr: {msg.header.stamp}")
+
+            self.prev_timestamp = msg.header.stamp
```

Comment:
I just realized that if we get an out of order stamp, we're still setting the prev timestamp to the current, which may be in the past, and if so by more than one step can allow a older-than-head message to sneak in. We should likely `return` once we discover the current message is out of order.

Reply:
Good catch!